File: app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistManager.java
package org.schabi.newpipe.local.playlist;

import androidx.annotation.Nullable;

import org.schabi.newpipe.database.AppDatabase;
import org.schabi.newpipe.database.playlist.PlaylistDuplicatesEntry;
import org.schabi.newpipe.database.playlist.PlaylistMetadataEntry;
import org.schabi.newpipe.database.playlist.PlaylistStreamEntry;
import org.schabi.newpipe.database.playlist.dao.PlaylistDAO;
import org.schabi.newpipe.database.playlist.dao.PlaylistStreamDAO;
import org.schabi.newpipe.database.playlist.model.PlaylistEntity;
import org.schabi.newpipe.database.playlist.model.PlaylistStreamEntity;
import org.schabi.newpipe.database.stream.dao.StreamDAO;
import org.schabi.newpipe.database.stream.model.StreamEntity;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class LocalPlaylistManager {
    private static final long THUMBNAIL_ID_LEAVE_UNCHANGED = -2;

    private final AppDatabase database;
    private final StreamDAO streamTable;
    private final PlaylistDAO playlistTable;
    private final PlaylistStreamDAO playlistStreamTable;

    public LocalPlaylistManager(final AppDatabase db) {
        database = db;
        streamTable = db.streamDAO();
        playlistTable = db.playlistDAO();
        playlistStreamTable = db.playlistStreamDAO();
    }

    public Maybe<List<Long>> createPlaylist(final String name, final List<StreamEntity> streams) {
        // Disallow creation of empty playlists
        if (streams.isEmpty()) {
            return Maybe.empty();
        }

        return Maybe.fromCallable(() -> database.runInTransaction(() -> {
                    final List<Long> streamIds = streamTable.upsertAll(streams);
                    final PlaylistEntity newPlaylist = new PlaylistEntity(name, false,
                            streamIds.get(0));

                    return insertJoinEntities(playlistTable.insert(newPlaylist),
                            streamIds, 0);
                }
        )).subscribeOn(Schedulers.io());
    }

    public Maybe<List<Long>> appendToPlaylist(final long playlistId,
                                              final List<StreamEntity> streams) {
        return playlistStreamTable.getMaximumIndexOf(playlistId)
                .firstElement()
                .map(maxJoinIndex -> database.runInTransaction(() -> {
                            final List<Long> streamIds = streamTable.upsertAll(streams);
                            return insertJoinEntities(playlistId, streamIds, maxJoinIndex + 1);
                        }
                )).subscribeOn(Schedulers.io());
    }

    private List<Long> insertJoinEntities(final long playlistId, final List<Long> streamIds,
                                          final int indexOffset) {

        final List<PlaylistStreamEntity> joinEntities = new ArrayList<>(streamIds.size());

        for (int index = 0; index < streamIds.size(); index++) {
            joinEntities.add(new PlaylistStreamEntity(playlistId, streamIds.get(index),
                    index + indexOffset));
        }
        return playlistStreamTable.insertAll(joinEntities);
    }

    public Completable updateJoin(final long playlistId, final List<Long> streamIds) {
        final List<PlaylistStreamEntity> joinEntities = new ArrayList<>(streamIds.size());
        for (int i = 0; i < streamIds.size(); i++) {
            joinEntities.add(new PlaylistStreamEntity(playlistId, streamIds.get(i), i));
        }

        return Completable.fromRunnable(() -> database.runInTransaction(() -> {
            playlistStreamTable.deleteBatch(playlistId);
            playlistStreamTable.insertAll(joinEntities);
        })).subscribeOn(Schedulers.io());
    }

    public Flowable<List<PlaylistMetadataEntry>> getPlaylists() {
        return playlistStreamTable.getPlaylistMetadata().subscribeOn(Schedulers.io());
    }

    public Flowable<List<PlaylistStreamEntry>> getDistinctPlaylistStreams(final long playlistId) {
        return playlistStreamTable
                .getStreamsWithoutDuplicates(playlistId).subscribeOn(Schedulers.io());
    }

    /**
     * Get playlists with attached information about how many times the provided stream is already
     * contained in each playlist.
     *
     * @param streamUrl the stream url for which to check for duplicates
     * @return a list of {@link PlaylistDuplicatesEntry}
     */
    public Flowable<List<PlaylistDuplicatesEntry>> getPlaylistDuplicates(final String streamUrl) {
        return playlistStreamTable.getPlaylistDuplicatesMetadata(streamUrl)
                .subscribeOn(Schedulers.io());
    }

    public Flowable<List<PlaylistStreamEntry>> getPlaylistStreams(final long playlistId) {
        return playlistStreamTable.getOrderedStreamsOf(playlistId).subscribeOn(Schedulers.io());
    }

    public Single<Integer> deletePlaylist(final long playlistId) {
        return Single.fromCallable(() -> playlistTable.deletePlaylist(playlistId))
                .subscribeOn(Schedulers.io());
    }

    public Maybe<Integer> renamePlaylist(final long playlistId, final String name) {
        return modifyPlaylist(playlistId, name, THUMBNAIL_ID_LEAVE_UNCHANGED, false);
    }

    public Maybe<Integer> changePlaylistThumbnail(final long playlistId,
                                                  final long thumbnailStreamId,
                                                  final boolean isPermanent) {
        return modifyPlaylist(playlistId, null, thumbnailStreamId, isPermanent);
    }

    public long getPlaylistThumbnailStreamId(final long playlistId) {
        return playlistTable.getPlaylist(playlistId).blockingFirst().get(0).getThumbnailStreamId();
    }

    public boolean getIsPlaylistThumbnailPermanent(final long playlistId) {
        return playlistTable.getPlaylist(playlistId).blockingFirst().get(0)
                .getIsThumbnailPermanent();
    }

    public long getAutomaticPlaylistThumbnailStreamId(final long playlistId) {
        final long streamId = playlistStreamTable.getAutomaticThumbnailStreamId(playlistId)
                .blockingFirst();
        if (streamId < 0) {
            return PlaylistEntity.DEFAULT_THUMBNAIL_ID;
        }
        return streamId;
    }

    private Maybe<Integer> modifyPlaylist(final long playlistId,
                                          @Nullable final String name,
                                          final long thumbnailStreamId,
                                          final boolean isPermanent) {
        return playlistTable.getPlaylist(playlistId)
                .firstElement()
                .filter(playlistEntities -> !playlistEntities.isEmpty())
                .map(playlistEntities -> {
                    final PlaylistEntity playlist = playlistEntities.get(0);
                    if (name != null) {
                        playlist.setName(name);
                    }
                    if (thumbnailStreamId != THUMBNAIL_ID_LEAVE_UNCHANGED) {
                        playlist.setThumbnailStreamId(thumbnailStreamId);
                        playlist.setIsThumbnailPermanent(isPermanent);
                    }
                    return playlistTable.update(playlist);
                }).subscribeOn(Schedulers.io());
    }

    public Maybe<Boolean> hasPlaylists() {
        return playlistTable.getCount()
                .firstElement()
                .map(count -> count > 0)
                .subscribeOn(Schedulers.io());
    }
}

File: app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistManager.kt
package org.schabi.newpipe.local.playlist

import androidx.room.withTransaction
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.firstOrNull
import org.schabi.newpipe.database.AppDatabase
import org.schabi.newpipe.database.playlist.PlaylistDuplicatesEntry
import org.schabi.newpipe.database.playlist.PlaylistMetadataEntry
import org.schabi.newpipe.database.playlist.PlaylistStreamEntry
import org.schabi.newpipe.database.playlist.model.PlaylistEntity
import org.schabi.newpipe.database.playlist.model.PlaylistStreamEntity
import org.schabi.newpipe.database.stream.model.StreamEntity

class LocalPlaylistManager(private val database: AppDatabase) {
    private val streamTable = database.streamDAO()
    private val playlistTable = database.playlistDAO()
    private val playlistStreamTable = database.playlistStreamDAO()

    companion object {
        private const val THUMBNAIL_ID_LEAVE_UNCHANGED = -2L
    }

    suspend fun createPlaylist(name: String, streams: List<StreamEntity>): List<Long>? {
        // Disallow creation of empty playlists
        if (streams.isEmpty()) {
            return null
        }

        // Save to the database directly.
        // Make sure the new playlist is always on the top of bookmark.
        // The index will be reassigned to non-negative number in BookmarkFragment.
        return database.withTransaction {
            val streamIds = streamTable.upsertAll(streams)
            val newPlaylist = PlaylistEntity(
                name = name,
                isThumbnailPermanent = false,
                thumbnailStreamId = streamIds[0],
                displayIndex = -1L
            )

            insertJoinEntities(
                playlistTable.insert(newPlaylist),
                streamIds,
                0
            )
        }
    }

    suspend fun appendToPlaylist(playlistId: Long, streams: List<StreamEntity>): List<Long>? {
        val maxJoinIndex = playlistStreamTable.getMaximumIndexOf(playlistId).firstOrNull() ?: -1
        return database.withTransaction {
            val streamIds = streamTable.upsertAll(streams)
            insertJoinEntities(playlistId, streamIds, maxJoinIndex + 1)
        }
    }

    private suspend fun insertJoinEntities(
        playlistId: Long,
        streamIds: List<Long>,
        indexOffset: Int
    ): List<Long> {
        val joinEntities = streamIds.mapIndexed { index, streamId ->
            PlaylistStreamEntity(playlistId, streamId, index + indexOffset)
        }
        return playlistStreamTable.insertAll(joinEntities)
    }

    suspend fun updateJoin(playlistId: Long, streamIds: List<Long>) {
        val joinEntities = streamIds.mapIndexed { index, streamId ->
            PlaylistStreamEntity(playlistId, streamId, index)
        }

        database.withTransaction {
            playlistStreamTable.deleteBatch(playlistId)
            playlistStreamTable.insertAll(joinEntities)
        }
    }

    suspend fun updatePlaylists(
        updateItems: List<PlaylistMetadataEntry>,
        deletedItems: List<Long>
    ) {
        val items = updateItems.map { PlaylistEntity(it) }
        database.withTransaction {
            for (uid in deletedItems) {
                playlistTable.deletePlaylist(uid)
            }
            for (item in items) {
                playlistTable.upsertPlaylist(item)
            }
        }
    }

    fun getDistinctPlaylistStreams(playlistId: Long): Flow<List<PlaylistStreamEntry>> {
        return playlistStreamTable.getStreamsWithoutDuplicates(playlistId)
    }

    /**
     * Get playlists with attached information about how many times the provided stream is already
     * contained in each playlist.
     *
     * @param streamUrl the stream url for which to check for duplicates
     * @return a list of {@link PlaylistDuplicatesEntry}
     */
    fun getPlaylistDuplicates(streamUrl: String): Flow<List<PlaylistDuplicatesEntry>> {
        return playlistStreamTable.getPlaylistDuplicatesMetadata(streamUrl)
    }

    fun getPlaylists(): Flow<List<PlaylistMetadataEntry>> {
        return playlistStreamTable.getPlaylistMetadata()
    }

    fun getPlaylistStreams(playlistId: Long): Flow<List<PlaylistStreamEntry>> {
        return playlistStreamTable.getOrderedStreamsOf(playlistId)
    }

    suspend fun renamePlaylist(playlistId: Long, name: String): Int? {
        return modifyPlaylist(playlistId, name, THUMBNAIL_ID_LEAVE_UNCHANGED, false)
    }

    suspend fun changePlaylistThumbnail(
        playlistId: Long,
        thumbnailStreamId: Long,
        isPermanent: Boolean
    ): Int? {
        return modifyPlaylist(playlistId, null, thumbnailStreamId, isPermanent)
    }

    suspend fun getPlaylistThumbnailStreamId(playlistId: Long): Long {
        return playlistTable.getPlaylist(playlistId).first()[0].thumbnailStreamId
    }

    suspend fun getIsPlaylistThumbnailPermanent(playlistId: Long): Boolean {
        return playlistTable.getPlaylist(playlistId).first()[0].isThumbnailPermanent
    }

    suspend fun getAutomaticPlaylistThumbnailStreamId(playlistId: Long): Long {
        val streamId = playlistStreamTable.getAutomaticThumbnailStreamId(playlistId).first().toLong()
        return if (streamId < 0) {
            PlaylistEntity.DEFAULT_THUMBNAIL_ID
        } else {
            streamId
        }
    }

    private suspend fun modifyPlaylist(
        playlistId: Long,
        name: String?,
        thumbnailStreamId: Long,
        isPermanent: Boolean
    ): Int? {
        val playlistEntities = playlistTable.getPlaylist(playlistId).firstOrNull()
        if (playlistEntities.isNullOrEmpty()) return null

        val playlist = playlistEntities[0]
        if (name != null) {
            playlist.name = name
        }
        if (thumbnailStreamId != THUMBNAIL_ID_LEAVE_UNCHANGED) {
            playlist.thumbnailStreamId = thumbnailStreamId
            playlist.isThumbnailPermanent = isPermanent
        }
        return playlistTable.update(playlist)
    }

    suspend fun hasPlaylists(): Boolean {
        val count = playlistTable.getCount().firstOrNull() ?: 0L
        return count > 0
    }
}
